keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/NotificationEventListener.kt

```diff
@@ -56,30 +67,156 @@ class NotificationEventListener(
+    with(notification) {
+      val deliveryArtifact = config.artifacts.find {
+        it.reference == veto.reference
+      } ?: return
+
+      val vetoedArtifact = repository.getArtifactVersion(deliveryArtifact, veto.version, null)
+      if (vetoedArtifact == null) {
+        log.warn("vetoedArtifact is null for application ${config.application}. Can't send MarkAsBadNotification notification")
+        return
+      }
+
+      val handler = handlers.find {
+        it.type == notification.type
+      } as? SlackNotificationHandler<SlackMarkAsBadNotification> ?: return
+
+      config.environments.forEach { environment ->
+        environment.notifications.filter {
+          it.type == NotificationType.slack
+        }.onEach {
+          handler.sendMessage(SlackMarkAsBadNotification(
+            vetoedArtifact = vetoedArtifact,
+            user = user,
+            targetEnvironment = veto.targetEnvironment,
+            channel = it.address,
+            time = clock.instant(),
+            application = config.name,
+            comment = veto.comment
+          ))
+        }
+      }
+    }
+  }
+
+  @EventListener(ApplicationActuationPaused::class)
+  fun onApplicationActuationPaused(notification: ApplicationActuationPaused) {
+    with(notification) {
+      val config = repository.getDeliveryConfigForApplication(application)
+
+      val handler = handlers.find {
+        it.type == Type.PAUSED_APPLICATION
+      } as? SlackNotificationHandler<SlackPausedNotification> ?: return
+
+      config.environments.forEach { environment ->
+        environment.notifications.filter {
+          it.type == NotificationType.slack
+        }.onEach {
+          handler.sendMessage(SlackPausedNotification(
+            user = triggeredBy,
+            channel = it.address,
+            time = clock.instant(),
+            application = application
+          ))
+        }
+      }
+    }
+  }
+
+  @EventListener(ApplicationActuationResumed::class)
+  fun onApplicationActuationResumed(notification: ApplicationActuationResumed) {
+    with(notification) {
+      val config = repository.getDeliveryConfigForApplication(application)
+
+      val handler = handlers.find {
+        it.type == Type.PAUSED_APPLICATION
+      } as? SlackNotificationHandler<SlackResumedNotification> ?: return
+
+      config.environments.forEach { environment ->
+        environment.notifications.filter {
+          it.type == NotificationType.slack
+        }.onEach {
+          handler.sendMessage(SlackResumedNotification(
+            user = triggeredBy,
+            channel = it.address,
+            time = clock.instant(),
+            application = application
+          ))
+        }
+      }
+    }
+  }
+
+  @EventListener(LifecycleEvent::class)
+  fun onLifecycleEvent(notification: LifecycleEvent) {
+    with(notification) {
+      val config = repository.getDeliveryConfig(artifactRef.split(":")[0])
```

Comment:
This logic of splitting the artifact reference and picking the right piece looks like a good candidate for functions/getters in `LifecyleEvent`.